subsys/net/lib/aws_fota/src/aws_fota.c

```diff
@@ -111,115 +111,192 @@ static int update_job_execution(struct mqtt_client *const client,
+ *
+ * @param[in] client Connected MQTT client instance
+ * @param[in] payload_len Length of the payload going to be read out from the
+ *			  MQTT message.
+ *
+ * @return 1 If successful otherwise a negative error code is returned.
+ */
+static int job_update_accepted(struct mqtt_client *const client,
+			       u32_t payload_len)
+{
+	int err = get_published_payload(client, payload_buf, payload_len);
 
-		/* Subscribe to update topic to recive feedback on wether an
-		 * update is accepted or not.
+	if (err) {
+		LOG_ERR("Error when getting the payload: %d", err);
+		return err;
+	}
+
+	/* Update accepted, so the execution version number needs to be
+	 * incremented. This can also be parsed out from the response but it's
+	 * better if the device handles this state so we need to send less data.
+	 * Also it means that the device is the driver of updates to this
+	 * document.
+	 */
+	execution_version_number++;
+
+	if (fota_state == DOWNLOAD_FIRMWARE) {
+		/*Job document is updated and we are ready to download
+		 * the firmware.
 		 */
-		err = aws_jobs_subscribe_topic_update(client, job_id,
-						      update_topic);
+		execution_state = AWS_JOBS_IN_PROGRESS;
+		LOG_INF("Start downloading firmware from %s/%s",
+			log_strdup(hostname), log_strdup(file_path));
+		err = fota_download_start(hostname, file_path);
 		if (err) {
-			LOG_ERR("Error when subscribing job_id_update: "
+			LOG_ERR("Error when trying to start firmware download: "
 				"%d", err);
 			return err;
 		}
+	} else if (execution_state == AWS_JOBS_IN_PROGRESS
+		   && fota_state == APPLY_UPDATE) {
+		LOG_INF("Firmware download completed");
+		execution_state = AWS_JOBS_SUCCEEDED;
+		err = update_job_execution(client, job_id,
+				execution_state, fota_state,
+				execution_version_number, "");
+		if (err) {
+			LOG_ERR("Unable to update the job execution");
+			return err;
+		}
+	} else if (execution_state == AWS_JOBS_SUCCEEDED &&
+			fota_state == APPLY_UPDATE) {
+		LOG_INF("Job document updated with SUCCEDED");
+		LOG_INF("Ready to reboot");
+		callback(AWS_FOTA_EVT_DONE);
+	}
+	return 1;
+}
 
-		/* Set fota_state to DOWNLOAD_FIRMWARE, when we are subscribed
-		 * to job_id topics we will try to publish and if accepted we
-		 * can start the download
-		 */
-		fota_state = DOWNLOAD_FIRMWARE;
+/**
+ * @brief Handling of a job document update when it is rejected.
+ *
+ * @param[in] client Connected MQTT client instance
+ * @param[in] payload_len Length of the payload going to be read out from the
+ *			  MQTT message.
+ *
+ * @return A negative error code is returned.
+ */
 
-		/* Handled by the library */
-		return 1;
+static int job_update_rejected(struct mqtt_client *const client,
+			       u32_t payload_len)
+{
+	LOG_ERR("Job document update was rejected");
+	int err = get_published_payload(client, payload_buf, payload_len);
 
-	} else if (aws_jobs_cmp(update_topic, topic, topic_len, "accepted")) {
+	if (err) {
+		LOG_ERR("Error when getting the payload: %d", err);
```

Comment:
I suggest "Error %d when getting the payload"

Reply:
Yeah, that's better thanks.